libraries/AP_Motors/AP_MotorsMatrixSlewRateLimited.cpp

```diff
@@ -0,0 +1,428 @@
+ * the PWM outputs, to limit the additional current draw during prop
+ * accelration. The slew rate was limited 6 PWM units (microseconds) per fast
+ * loop iteration (400Hz). 3DR probably felt that this workaround would prevent
+ * any communication failures, so they disabled the previous failure detection /
+ * recovery code.
+ *
+ * A few months later, 3DR decided to renable the motor recovery code, so the
+ * "final" workaround was to have both slew rate limiting and motor recovery
+ * enabled.
+ *
+ * The important parameters from 3DR's original implementation are:
+ *
+ *    Max slew rate: +- 6 PWM units (microseconds) / main loop iteration
+ *      - With the main loop of 400Hz, this is 2400 PWM units / second.
+ *
+ *    Initial Motor recovery output level: 50% * (max pwm - min pwm) + min pwm
+ *        Recovery output is percentage of difference between min / max pwm.
+ *
+ *    Motor recovery time: 250 milliseconds to go from min pwm to max pwm
+ *        As the motor recovery started at 50%, the actual recovery time is 125ms.
+ *        This recovery time is equivalent to a slew rate of 3600 PWM units / second.
+ *
+ * Note that all of these parameters are based on PWM values, and do not take
+ * into account thrust curves / limits / etc.
+ *
+ * All of the recovery / slew rate code is implemented in the final stage of the
+ * motor output pipeline. The "requested" output level is filtered by the
+ * reovery code, and the filtered result is the output level actually sent to
+ * the motors.
+ *
+ * When the motor is in a recovery state, the recovery code enforces a linearly
+ * increasing (with time) constraint on the maximum PWM output passed through to
+ * the motors. The recovery code does not limit the slew rate, just the output.
+ * In theory, if the requested output level was 50% of the max, and then at the
+ * last moment of recovery, increased to 100%, the actual PWM output would
+ * immediately increase from 50% to 100%. However, the input to the motor
+ * recovery filter is slew rate limited, so in practice, the rate of change in
+ * the output is going to fall between 2400 PWM units / second, and 3600 PWM
+ * units / second.
+ *
+ * AP_MotorsMatrixSlewRateLimited is a rewrite of 3DR's original slew rate
+ * limiting, motor failure detection, and motor recovery code. The main
+ * difference between this implementation and the original is that the slew rate
+ * limiter in this implementation also handles the soft restart for motor
+ * recovery. In the original code, these functions were handled differently.
+ */
+
+#include <AP_HAL/AP_HAL.h>
+#include "AP_MotorsMatrixSlewRateLimited.h"
+
+// parameters for the motor class
+const AP_Param::GroupInfo AP_MotorsMatrixSlewRateLimited::var_info[] = {
+    AP_NESTEDGROUPINFO(AP_MotorsMulticopter, 0),
+
+    // @Param: SLW_RT_NORM
+    // @DisplayName: Normal Motor Slew Rate
+    // @Description: Maximum allowed change in motor PWM output per second
+    // @Units: pwm/sec
```

Comment:
To be consistent, this must be PWM/s and it must be added to Tools/autotest/param_metadata/param.py around line 48